File: karate_chop/python/chop.py
def chop(target, field):
  if len(field) == 0 :
    return -1
  elif len(field) == 1:
    if field[0] == target:
      return 0
    else:
      return -1
  else:
    m = len(field)//2
    if target < field[m]:
      return chop(target, field[:m])
    else:
      index = chop(target, field[m:])
      if index == -1:
        return -1
      else:
        return index + m

File: karate_chop/cpp/chop.cc
#include "chop.h"
#include <vector>

int chop(int n, std::vector<int> field) {
  if (field.size() == 0) {
    return -1;
  }

  int begin = 0;
  int end = field.size() - 1;
  
  while (begin != end) {
    int middle = (end - begin) / 2;
    if ( n <= field[middle] ){
      end = middle;
    } else {
      begin = middle + 1;
    }
  }

  if ( field[begin] == n ) {
    return begin;
  } else {
    return -1;
  }
}
